Feature: allow DIP upload to Binder

jrwdunham opened the pull request:
- Adds `access_system_id` attribute to the `Transfer` model (including migration).
- Adds "Access System ID" input to transfer browser GUI.
- Adds `access_system_id` param to payload of POST v2beta/package endpoint.
- Fixes dependency of access_system_id migration.
- Updates code comments and HTML help text to explain how to use the "Access System ID" at the start of transfer or the AtoM/Binder Identifier in the modal dialog to specify the identifier required by the AtoM or Binder SWORD V1 deposit endpoint during DIP upload to AtoM or Binder.

Connected to archivematica/Issues#23

**Note 1**: this PR was built by first cherry-picking the following four commits from the `dev/1.6.x-moma` branch and then adding a new commit:

- fa54b3014cb63f0a656d37624f31d93137a8a43f
- 3a7e56cb8d57a828233ac7c1c486e1f9300e3b06
- 15cb8d1eacdf821850517d18959e380b1075e290
- 0fa2fb4392f8e8ac54a8939a83d875db96bfbbe9

**Note 2**: this work has NOT yet been tested with DIP upload to a real AtoM or Binder instance.

The transfer browser should now look like the following:

<img width="1013" alt="screen shot 2018-07-20 at 9 40 00 am" src="https://user-images.githubusercontent.com/2504271/43026017-bf22fc82-8c28-11e8-9751-b78b9638d61a.png">

The modal dialog in ingest at the "Choose config for AtoM DIP upload" micro-service should now look like the following:

<img width="608" alt="screen shot 2018-07-20 at 2 28 41 pm" src="https://user-images.githubusercontent.com/2504271/43026239-afef1772-8c29-11e8-8cee-39464212b686.png">

mcantelon committed:
JS: automatically fill in AtoM upload form

commit 2d7efa7fcdcad0f5a337a5e41b884c6079decf87
Author: Mike Cantelon <mcantelon@gmail.com>
Date:   Wed Mar 19 23:46:30 2014 -0700

    Fixed issues with DRMC DIP target prefix selection.

commit 5bdee041818446d05832299619c26a30bb9f1ce4
Author: Mike Cantelon <mcantelon@gmail.com>
Date:   Wed Mar 19 16:22:41 2014 -0700

    Added DIP target prefix selection.

commit 884b8083524f96a15783d8cbd141c14e4567f4ef
Author: Mike Cantelon <mcantelon@gmail.com>
Date:   Wed Mar 12 14:00:46 2014 -0700

    Added use of access # for AtoM DIP slug.

commit 7c18cf8b2fbf2b08dcfba4a18346d68b629cef2f
Author: Mike Cantelon <mcantelon@gmail.com>
Date:   Wed Mar 5 00:38:38 2014 -0800

    Only ask for DIP upload slug if no accession number.

Hwesta committed:
Transfer: Store access system ID

Allow an access system ID to be submitted when a transfer is created. This
can be used later to help automate submitting a DIP to an access system.

Hwesta committed:
Cleanup

Hwesta committed:
Upload to AtoM: Use access_system_id instead of accession no

When uploading a DIP to AtoM, use the new access_system_id instead of the
overloaded accession no.

sevein commented on src/dashboard/frontend/transfer-browser/app/front_page/content.html:
```diff
@@ -22,6 +22,11 @@
 
+  <div class="col-xs-2">
+  <input class="form-control" ng-model="vm.transfer.access_system_id"/>
+  <div class="help-block">{{ "Access System ID" | translate }}</div>
```
Other labels are only sentence cased. Should be `Access system ID` instead?

jrwdunham replied:
k.

sevein commented on src/dashboard/src/templates/ingest/grid.html:
```diff
@@ -246,13 +246,16 @@ <h3>{% trans "Edit metadata" %}</h3>
+          <p>{% trans "Please provide the identifier required by the AtoM or Binder SWORD V1 deposit endpoint." %}</p>
+          <ul>
+              <li><strong>AtoM:</strong>&nbsp;{% trans "In AtoM, this is the permalink or slug of the target archival description where the DIP will be attached as a child. For example, if the URL of the archival description is 'http://myAtoM.ca/newsletters-2;rad', then you would enter 'newsletters-2'." %}</li>
+              <li><strong>Binder:</strong>&nbsp;{% trans "In Binder, this is an identifier that indicates which artwork or technology record the DIP will be attached to. For example, if the DIP is being uploaded to an artwork record, and that artwork has a TMS Object ID of 123, you would enter 'ar:123'. If the DIP is being uploaded to a technology record, and that technology record has a URL in B ...
```
Two observations:
- Our templates use two spaces for indentation.
- I think that it's preferably to annotate the whole sentence, e.g. `AtoM: ...` even if it needs to contains HTML elements that way we don't enforce a grammar structure.

sevein commented on src/dashboard/src/main/migrations/0055_transfer_access_system_id.py:
```diff
@@ -7,7 +7,7 @@
 
     dependencies = [
-        ('main', '0029_clean_transfers'),
+        ('main', '0054_index_aip_silent_error_branch'),
```
I've just added 0055 (#1216).

sevein reviewed (approved):
LGTM.

sromkey commented:
I have a question and a suggestion:
Question: in the user interface at DIP upload, has that text changed? If Binder isn't split out into its own DIP upload selection in the drop-down menu we should ammend the text to read "Upload DIP to AtoM/Binder."
Suggestion: In the modal dialog, to make the Binder text a little less art-repository-specific and to generalize away the use of TMS (it's a longer-term goal that Binder shouldn't rely on a connected TMS instance to work). Suggested text:

`Binder: In Binder, this is an identifier that indicates which resource or technology record the DIP will be attached to. Resources are preceded by 'ar:' (e.g. 'ar:123') and technology records are preceded by 'tr:' (e.g. 'tr:456')`

Further explanation about how to find the appropriate id numbers for resource and technology records should be covered in documentation.

jrwdunham commented:
@sevein @sromkey I have made the changes you suggested.

@sromkey here's what the UI currently looks like (in this PR) wrt AtoM/Binder:

- On the processing config page (e.g., /administration/processing/edit/default/), the decision says "Upload DIP to AtoM".
- On the ingest page under "Upload DIP", the relevant micro-service name is "Choose config for AtoM DIP upload" with the only option in the select box being "- upload-qubit_v0.0".

Should these ^ be changed to reference Binder as well?

sromkey commented:
@jrwdunham-

- Yeah, I think the Processing Config page should also refer to Binder, good catch
- The Choose Config for AtoM DIP upload is not actually something a user ever sees, I'm talking about one step before that- choosing between DIP upload methods (e.g. ArchivesSpace, CONTENTdm, AtoM, etc)

ablwr commented:
+1 to @sromkey's suggestions of directing the user to upload DIP to AtoM/Binder and the non-TMS-specific language. Which you've already done soooo.... LGTM!!! 👍

jrwdunham committed:
AtoM/Binder DIP upload: add access system ID

- Adds "Access System ID" input to transfer browser GUI.
- Adds ``access_system_id`` param to payload of POST v2beta/package
  endpoint.
- Fixes dependency of access_system_id migration.
- Updates code comments and HTML help text to explain how to use the
  "Access System ID" at the start of transfer or the AtoM/Binder
  Identifier in the modal dialog to specify the identifier required by the AtoM
  or Binder SWORD V1 deposit endpoint during DIP upload to AtoM or
  Binder.
